dev/com.ibm.ws.microprofile.health/src/com/ibm/ws/microprofile/health/internal/AppTrackerImpl.java

```diff
@@ -238,6 +253,70 @@ public boolean isStarted(String appName) {
         }
     }
 
+    /**
+     * Returns true if the application with the specified name is installed, otherwise false.
+     *
+     * @return true if the application with the specified name is installed, otherwise false.
+     */
+    @Override
+    public boolean isInstalled(String appName) {
+        lock.readLock().lock();
+        try {
+            String state = getApplicationMBean(appName);
+            if (state.isEmpty()) {
+                appStateMap.replace(appName, null);
+            }
+            else if (state.equals("INSTALLED")) {
+                if (appStateMap.get(appName) != ApplicationState.INSTALLED) {
+                    appStateMap.replace(appName, ApplicationState.INSTALLED);
+                }
+                return true;
+            }
+        } finally {
+            lock.readLock().unlock();
+        }
+        return false;
+    }
+
+    /**
+     * Returns true if the application with the specified name is uninstalled, otherwise false.
+     *
+     * @return true if the application with the specified name is uninstalled, otherwise false.
+     */
+    @Override
+    public boolean isUninstalled(String appName) {
+        lock.readLock().lock();
+        try {
+            if (appStateMap.get(appName) == null) {
+                return true;
+            }
+        } finally {
+            lock.readLock().unlock();
+        }
+        return false;
+    }
+
+    /**
+     * Returns the MBeanInfo of appName if the ApplicationMBean exists, otherwise null.
+     *
+     * @return the MBeanInfo of appName if the ApplicationMBean exists, otherwise null.
+     */
+    private String getApplicationMBean(String appName) {
+        lock.readLock().lock();
```

Comment:
Do we need a lock here? I think there maybe scenarios where the lock is already being held from isInstalled() or its caller, and it will never get into this method, leading to a potential deadlock.

Reply:
Makes sense, I have removed the lock.